validator/keymanager/remote-web3signer/keymanager.go

```diff
@@ -38,54 +48,249 @@ type SetupConfig struct {
+		}
+		if len(line) != pubkeyLength {
+			log.WithFields(log.Fields{
+				"filepath": km.keyFilePath,
+				"key":      line,
+			}).Fatal("Invalid public key in remote signer key file")
+			continue
+		}
+		if _, found := seenKeys[line]; !found {
+			// If it's a new line, mark it as seen and process it
+			pubkey, err := hexutil.Decode(line)
+			if err != nil {
+				return nil, nil, errors.Wrapf(err, "could not decode public key line in remote signer key file: %s", line)
+			}
+			bPubkey := bytesutil.ToBytes48(pubkey)
+			seenKeys[line] = bPubkey
+			keys = append(keys, bPubkey)
+		}
+	}
+	// Check for scanning errors
+	if err := scanner.Err(); err != nil {
+		return nil, nil, errors.Wrap(err, "could not scan remote signer public key file")
+	}
+	if len(keys) == 0 {
+		log.Warn("Remote signer key file: no valid public keys found")
+	}
+	return keys, seenKeys, nil
+}
+
+func (km *Keymanager) savePublicKeysToFile(providedPublicKeys map[string][48]byte) error {
+	if km.keyFilePath == "" {
+		return errors.New("no key file provided")
+	}
+	pubkeys := make([][48]byte, 0)
+	var bytesBuf bytes.Buffer
+	for key, value := range providedPublicKeys {
+		bytesBuf.WriteString(key)
+		bytesBuf.WriteString("\n")
+		pubkeys = append(pubkeys, value)
+	}
+
+	km.lock.Lock()
+	if err := file.WriteFile(km.keyFilePath, bytesBuf.Bytes()); err != nil {
+		km.lock.Unlock()
+		return errors.Wrapf(err, "could not write to file %s", km.keyFilePath)
+	}
+	km.providedPublicKeys = pubkeys
+	km.lock.Unlock()
+	return nil
+}
+
+func (km *Keymanager) refreshRemoteKeysFromFileChanges(ctx context.Context) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		log.WithError(err).Error("Could not initialize file watcher")
+		return
+	}
+	defer func() {
+		if err := watcher.Close(); err != nil {
+			log.WithError(err).Error("Could not close file watcher")
+		}
+	}()
+	if err := watcher.Add(km.keyFilePath); err != nil {
+		log.WithError(err).WithField("filepath", km.keyFilePath).Errorf("Could not add file to file watcher")
+		return
+	}
+	for {
+		select {
+		case e, ok := <-watcher.Events:
+			if !ok { // Channel was closed (i.e. Watcher.Close() was called).
+				return
+			}
+			if e.Op.String() == "REMOVE" {
+				log.Fatalln("Remote signer keyfile was removed! Restart the validator client with the appropriate remote signer file")
+			}
+			log.WithFields(log.Fields{
+				"event": e.Name,
+				"op":    e.Op.String(),
+			}).Info("remote signer keyfile event triggered")
+			fileKeys, _, err := km.readKeyFile()
+			if err != nil {
+				log.WithError(err).Fatalln("Could not read key file")
+			}
+			if len(fileKeys) == 0 {
+				log.Warnln("Remote signer keyfile no longer has keys, defaulting to flag provided keys")
+				fileKeys = km.flagLoadedPublicKeys
+			}
+			currentKeys, err := km.FetchValidatingPublicKeys(ctx)
+			if err != nil {
+				log.WithError(err).Fatalln("Could not fetch current keys")
+			}
+			if !reflect.DeepEqual(currentKeys, fileKeys) {
```

Comment:
is there a better way of handling this?

Reply:
https://pkg.go.dev/golang.org/x/exp/slices#Equal

Did you try slices.Equal? Since this is a type of [][48]byte, it should work.

Reply:
can try that

Reply:
I think that worked thanks